### PR TITLE
Fix like function when no escape char specified

### PR DESCRIPTION
**Pull request by aditi-pandit**
The 'like' function logic had a bug when escape char
was not specified in its invocation. In this situation,
the escape char was left uninitialized and could pick up
any value. If that random value matched a char in the
'pattern' parameter then it resulted in failures with error
'Escape character must be followed by ...'

This change fixes the bug by tracking using a boolean if indeed the
escape char was specified in the invocation, and modifying the logic
accordingly.

### CODE REVIEW

File: velox/functions/lib/Re2Functions.cpp [hunk 1]
```diff
@@ -125,18 +125,21 @@ bool re2Extract(
   }
 }
 
-std::string
-likePatternToRe2(StringView pattern, char escapeChar, bool& validPattern) {
+std::string likePatternToRe2(
+    StringView pattern,
+    bool hasEscape,
```
Comment (mbasmanova) [inline velox/functions/lib/Re2Functions.cpp:130]: nit: using std::optional for escapeChar might be clearer

File: velox/functions/lib/Re2Functions.cpp [hunk 2]
```diff
@@ -738,6 +745,7 @@ std::shared_ptr<exec::VectorFunction> makeLike(
       name,
       inputArgs[1].type->toString());
 
+  bool hasEscape = false;
```
Comment (mbasmanova) [inline velox/functions/lib/Re2Functions.cpp:748]: nit: setting `hasEscape = numArgs == 3` and using `if (hasEscape)` might be clearer

File: velox/functions/lib/Re2Functions.cpp (hunk 1 above)
Comment (funrollloops) [reply velox/functions/lib/Re2Functions.cpp:130]: You sure? The code becomes a lot cleaner elsewhere (see other comments).

File: velox/functions/lib/Re2Functions.cpp [hunk 3]
```diff
@@ -125,18 +125,21 @@ bool re2Extract(
...
+    char escapeChar,
+    bool& validPattern) {
   std::string regex;
   validPattern = true;
   regex.reserve(pattern.size() * 2);
   regex.append("^");
   bool escaped = false;
   for (const char c : pattern) {
-    if (escaped && !(c == '%' || c == '_' || c == escapeChar)) {
+    if (hasEscape && escaped && !(c == '%' || c == '_' || c == escapeChar)) {
```
Comment (funrollloops) [inline velox/functions/lib/Re2Functions.cpp:139]: If you used a `std::optional` for `escapeChar`, no changes would be needed to the body of this function. In particular these conditions would be noticeably simpler IMO.

File: velox/functions/lib/Re2Functions.cpp [hunk 4]
```diff
@@ -738,8 +744,9 @@ std::shared_ptr<exec::VectorFunction> makeLike(
       name,
       inputArgs[1].type->toString());
 
+  bool hasEscape = numArgs == 3;
   char escapeChar;
```
Comment (funrollloops) [inline velox/functions/lib/Re2Functions.cpp:748]: If you used a `std::optional` here, you wouldn't need any other changes to this function.

**Pull request closed without merging**